File: Api/src/main/java/net/winternetwork/bedwars/api/score/ReplaceableList.java
package net.winternetwork.bedwars.api.score;

import lombok.RequiredArgsConstructor;
import org.bukkit.entity.Player;

import java.util.List;
import java.util.function.BiFunction;
import java.util.stream.Collectors;

@RequiredArgsConstructor
public class ReplaceableList {

    private final List<String> list;
    private final BiFunction<String, Player, String> function;

    public List<String> call(Player player) {
        return list.stream()
                .map(it -> function.apply(it, player))
                .collect(Collectors.toList());
    }
}

File: Api/src/main/kotlin/net/winternetwork/bedwars/api/score/ReplaceableList.kt
package net.winternetwork.bedwars.api.score

import org.bukkit.entity.Player

//
//class ReplaceableList(
//        private val list: List<String>,
//        private val function: (String, Player) -> String
//) {
//    fun call(player: Player): List<String> {
//        return list.map {
//            function.invoke(it, player)
//        }
//    }
//}


class ReplaceableList(
    private val list: List<String>,
    private val replaces: (Player) -> Map<String, String>
) {

    fun call(player: Player): List<String> {
        return list.map { text ->
            text.apply {
                replaces(player).forEach { (key, value) ->
                    this.replace(key, value)
                }
            }
        }
    }
}